Database unavailable

fbiville opened the pull request.

fbiville committed:
Improve router tests covering dead connections

The previous setup relies on a specific test implementation of
the callback instead of just verifying that the callback is called.

bigmontz commented on neo4j/internal/retry/state.go:
```diff
@@ -134,6 +136,19 @@ func (s *State) OnFailure(ctx context.Context, conn idb.Connection, err error, i
 			s.cause = "Cluster error"
 			return
 		}
+		if dbErr.Code == "Neo.TransientError.General.DatabaseUnavailable" {
+			// invalidate server since it does not host the requested db
+			invalidationFn := s.Router.InvalidateWriter
+			if mode == idb.ReadMode {
+				invalidationFn = s.Router.InvalidateReader
+			}
+			if err := invalidationFn(ctx, s.DatabaseName, conn.ServerName()); err != nil {
+				s.stop = true
+				s.LastErr = err
+			}
+			s.cause = "Database unavailable error"
+			return
```
if a `conn` to the router arrives here, should it be remove from write/read or just invalidate the server name?

fbiville replied:
can you get a `Neo.TransientError.General.DatabaseUnavailable` with a router connection? 🤔

fbiville replied:
moreover, this snippet here is called when transaction function failures occur, so a router connection would not be used here, but rather a reader or a writer one, or?

fbiville committed:
Mark DB unavailable error as fatal

In turn, this will purge all the server's connections upon the
return of the failed connection to the pool.
